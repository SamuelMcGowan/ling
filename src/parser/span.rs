use std::ops::Range;

#[derive(Clone, Copy, PartialEq, Eq)]
pub struct Span {
    start: usize,
    end: usize,
}

impl Span {
    #[inline]
    pub fn new(start: usize, end: usize) -> Self {
        Self {
            start,
            end: end.max(start),
        }
    }

    #[inline]
    pub fn start(&self) -> usize {
        self.start
    }

    #[inline]
    pub fn end(&self) -> usize {
        self.end
    }

    #[inline]
    pub fn len(&self) -> usize {
        self.end - self.start
    }

    #[inline]
    pub fn is_empty(&self) -> bool {
        self.start == self.end
    }

    #[inline]
    pub fn range(&self) -> Range<usize> {
        self.start..self.end
    }

    #[inline]
    pub fn union(&self, other: Self) -> Self {
        Self {
            start: self.start.min(other.start),
            end: self.end.max(other.end),
        }
    }

    #[inline]
    pub fn intersection(&self, other: Self) -> Option<Self> {
        let start = self.start.max(other.start);
        let end = self.end.min(other.end);

        if start < end {
            Some(Self { start, end })
        } else {
            None
        }
    }
}

impl std::fmt::Debug for Span {
    #[inline]
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.range().fmt(f)
    }
}
