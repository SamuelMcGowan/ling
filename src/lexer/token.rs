use serde::Serialize;

use crate::constants::ConstIdx;
use crate::syntax::source::Span;

#[derive(Debug, Clone, Copy, Serialize)]
pub(crate) struct Token {
    pub span: Span,
    pub kind: TokenKind,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
pub(crate) enum TokenKind {
    Punct(Punct),
    Bracket(Bracket),

    Keyword(Keyword),

    Ident(Ustr),
    Const(ConstIdx),

    Error(Error),
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
pub(crate) enum Punct {
    Dot,
    Comma,
    Colon,
    Semicolon,

    Add,
    Sub,
    Mul,
    Div,
    Mod,
    Pow,

    AddEqual,
    SubEqual,
    MulEqual,
    DivEqual,
    ModEqual,
    PowEqual,

    Gt,
    Lt,
    Equal,

    GtEqual,
    LtEqual,
    EqualEqual,

    BangEqual,
    Bang,

    FatArrow,
    Arrow,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
pub(crate) enum Bracket {
    Opening(BracketKind),
    Closing(BracketKind),
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
pub(crate) enum BracketKind {
    Round,
    Curly,
    Square,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
pub(crate) enum Keyword {
    Func,
    Return,

    Type,
    Enum,
    Data,
    Typeclass,
    Impl,

    Let,

    Match,
    Is,

    LogicalAnd,
    LogicalOr,
    LogicalNot,

    If,
    Elif,
    Else,

    Loop,
    While,

    Unit,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
pub(crate) enum Error {
    UnexpectedChar(char),

    InvalidEscape,
    UnterminatedString,

    IntegerOverflow,
    MissingDigits,
}

macro_rules! tkind {
    (punct $punct:ident) => {{
        use crate::lexer::token::*;
        TokenKind::Punct(Punct::$punct)
    }};
    (bracket $side:ident $bracket:ident) => {{
        use crate::lexer::token::*;
        TokenKind::Bracket(Bracket::$side(BracketKind::$bracket))
    }};

    (kwd $kwd:ident) => {{
        use crate::lexer::token::*;
        TokenKind::Keyword(Keyword::$kwd)
    }};

    (ident $ident:expr) => {{
        use ustr::Ustr;
        use crate::lexer::token::*;

        TokenKind::Ident(Ustr::from($ident))
    }};
    (constant $constant:literal) => {{
        use crate::lexer::token::*;
        use crate::constants::ConstIdx;

        TokenKind::Const(ConstIdx($constant))
    }};

    (error $error:ident $($value:tt)?) => {{
        use crate::lexer::token::*;
        TokenKind::Error(Error::$error $($value)?)
    }}
}

pub(crate) use tkind;
use ustr::Ustr;
