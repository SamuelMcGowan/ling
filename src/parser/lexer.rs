use super::span::Span;
use super::token::{tkind, Token};
use super::ParseContext;

impl<'a> ParseContext<'a> {
    pub fn lex_token(&mut self) -> Option<Token> {
        loop {
            let c = self.cursor.next()?;
            let start_pos = self.cursor.byte_pos();

            let kind = match c {
                c if c.is_ascii_whitespace() => continue,
                '#' => {
                    while !matches!(self.cursor.next(), Some('\n') | None) {}
                    continue;
                }

                '(' => tkind!(punct LParen),
                ')' => tkind!(punct RParen),
                '{' => tkind!(punct LBrace),
                '}' => tkind!(punct RBrace),
                '[' => tkind!(punct LBracket),
                ']' => tkind!(punct RBracket),

                '.' => tkind!(punct Dot),
                ',' => tkind!(punct Comma),
                ':' => tkind!(punct Colon),
                ';' => tkind!(punct Semicolon),

                '-' if self.cursor.eat('>') => tkind!(punct Arrow),
                '=' if self.cursor.eat('>') => tkind!(punct FatArrow),

                '+' if self.cursor.eat('=') => tkind!(punct AddEqual),
                '-' if self.cursor.eat('=') => tkind!(punct SubEqual),
                '*' if self.cursor.eat('=') => tkind!(punct MulEqual),
                '/' if self.cursor.eat('=') => tkind!(punct DivEqual),
                '%' if self.cursor.eat('=') => tkind!(punct ModEqual),
                '^' if self.cursor.eat('=') => tkind!(punct PowEqual),

                '+' => tkind!(punct Add),
                '-' => tkind!(punct Sub),
                '*' => tkind!(punct Mul),
                '/' => tkind!(punct Div),
                '%' => tkind!(punct Mod),
                '^' => tkind!(punct Pow),

                '>' if self.cursor.eat('=') => tkind!(punct GtEqual),
                '<' if self.cursor.eat('=') => tkind!(punct LtEqual),
                '=' if self.cursor.eat('=') => tkind!(punct EqualEqual),

                '>' => tkind!(punct Gt),
                '<' => tkind!(punct Lt),
                '=' => tkind!(punct Equal),

                _ => tkind!(error UnexpectedChar(c)),
            };

            let span = Span::new(start_pos, self.cursor.byte_pos());

            return Some(Token { span, kind });
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::parser::token::TokenKind;

    #[test]
    fn punct_single() {
        check_tokens(
            "+-:;",
            &[
                tkind!(punct Add),
                tkind!(punct Sub),
                tkind!(punct Colon),
                tkind!(punct Semicolon),
            ],
        );
    }

    #[test]
    fn punct_double() {
        check_tokens(
            "->==+=",
            &[
                tkind!(punct Arrow),
                tkind!(punct EqualEqual),
                tkind!(punct AddEqual),
            ],
        )
    }

    #[test]
    fn whitespace() {
        check_tokens(" .\r\n. ", &[tkind!(punct Dot), tkind!(punct Dot)])
    }

    #[test]
    fn comment() {
        check_tokens(".# comment\r\n.# comment", &[tkind!(punct Dot), tkind!(punct Dot)]);
    }

    #[test]
    fn comment_empty() {
        check_tokens(".#\r\n.#", &[tkind!(punct Dot), tkind!(punct Dot)])
    }

    fn check_tokens(s: &str, t: &[TokenKind]) {
        let mut context = ParseContext::new(s);

        let mut tokens = vec![];
        while let Some(token) = context.lex_token() {
            tokens.push(token.kind);
        }

        assert_eq!(&tokens, t)
    }
}
