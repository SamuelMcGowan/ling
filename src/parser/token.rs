use super::span::Span;

#[derive(Debug, Clone, Copy)]
pub struct Token {
    pub span: Span,
    pub kind: TokenKind,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TokenKind {
    Punct(Punct),
    Keyword(Keyword),
    Ident(Ustr),
    Error(Error),
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Punct {
    LParen,
    RParen,
    LBrace,
    RBrace,
    LBracket,
    RBracket,

    Dot,
    Comma,
    Colon,
    Semicolon,

    Add,
    Sub,
    Mul,
    Div,
    Mod,
    Pow,

    AddEqual,
    SubEqual,
    MulEqual,
    DivEqual,
    ModEqual,
    PowEqual,

    Gt,
    Lt,
    Equal,

    GtEqual,
    LtEqual,
    EqualEqual,

    FatArrow,
    Arrow,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Keyword {
    Func,
    Return,

    Type,
    Enum,
    Data,
    Typeclass,
    Impl,

    Let,
    Match,

    If,
    Elif,
    Else,

    Loop,
    While,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Error {
    UnexpectedChar(char),
}

macro_rules! tkind {
    (punct $punct:ident) => {{
        use crate::parser::token::*;
        TokenKind::Punct(Punct::$punct)
    }};
    (kwd $kwd:ident) => {{
        use crate::parser::token::*;
        TokenKind::Keyword(Keyword::$kwd)
    }};
    (ident $ident:expr) => {{
        use ustr::Ustr;
        use crate::parser::token::*;

        TokenKind::Ident(Ustr::from($ident))
    }};
    (error $error:ident $($value:tt)?) => {{
        use crate::parser::token::*;
        TokenKind::Error(Error::$error $($value)?)
    }}
}

pub(crate) use tkind;
use ustr::Ustr;
