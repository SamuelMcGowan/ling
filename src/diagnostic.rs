use codespan_reporting::diagnostic::{Diagnostic, Severity};
use codespan_reporting::term;
#[cfg(test)]
use codespan_reporting::term::termcolor::NoColor;
#[cfg(not(test))]
use codespan_reporting::term::termcolor::{ColorChoice, StandardStream};
#[cfg(test)]
use serde::Serialize;

use crate::lexer::token::Bracket;
use crate::source::{SourceDb, Span};

pub(crate) struct DiagnosticOutput {
    #[cfg(not(test))]
    output: StandardStream,

    #[cfg(test)]
    output: NoColor<Vec<u8>>,

    config: term::Config,
    had_errors: bool,
}

impl Default for DiagnosticOutput {
    fn default() -> Self {
        Self {
            #[cfg(not(test))]
            output: StandardStream::stderr(ColorChoice::Auto),

            #[cfg(test)]
            output: NoColor::new(vec![]),

            config: term::Config::default(),
            had_errors: false,
        }
    }
}

#[cfg(test)]
impl Serialize for DiagnosticOutput {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        let s = String::from_utf8_lossy(self.output.get_ref());
        s.serialize(serializer)
    }
}

impl DiagnosticOutput {
    pub fn reporter<'a>(
        &'a mut self,
        source_db: &'a SourceDb,
        source_id: usize,
    ) -> DiagnosticReporter {
        DiagnosticReporter {
            output: self,
            source_db,
            source_id,
        }
    }

    pub fn had_errors(&self) -> bool {
        self.had_errors
    }
}

pub(crate) struct DiagnosticReporter<'a> {
    output: &'a mut DiagnosticOutput,

    source_db: &'a SourceDb,
    source_id: usize,
}

impl<'a> DiagnosticReporter<'a> {
    pub fn borrow(&mut self) -> DiagnosticReporter {
        DiagnosticReporter {
            source_db: self.source_db,
            output: self.output,
            source_id: self.source_id,
        }
    }

    pub fn report(&mut self, diagnostic: impl IntoDiagnostic) {
        let diagnostic = diagnostic.into_diagnostic(self.source_id);
        self.output.had_errors |= diagnostic.severity >= Severity::Error;

        term::emit(
            &mut self.output.output,
            &self.output.config,
            self.source_db,
            &diagnostic,
        )
        .expect("failed to emit diagnostic");
    }
}

pub trait IntoDiagnostic {
    fn into_diagnostic(self, file_id: usize) -> Diagnostic<usize>;
}

// TODO: implement `Display` for brackets and tokens.

pub(crate) struct BracketMismatch {
    pub bracket: Bracket,
    pub span: Span,
}

impl IntoDiagnostic for BracketMismatch {
    fn into_diagnostic(self, file_id: usize) -> Diagnostic<usize> {
        Diagnostic::error()
            .with_labels(vec![self.span.with_source(file_id).label_primary()])
            .with_message(format!("mismatched {:?}", self.bracket))
    }
}
