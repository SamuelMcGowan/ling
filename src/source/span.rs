use std::ops::Range;

use codespan_reporting::diagnostic::Label;

#[derive(Clone, Copy, PartialEq, Eq, serde::Serialize)]
pub(crate) struct Span {
    start: usize,
    end: usize,
}

impl Span {
    pub fn new(start: usize, end: usize) -> Self {
        Self {
            start,
            end: end.max(start),
        }
    }

    pub fn start(&self) -> usize {
        self.start
    }

    pub fn end(&self) -> usize {
        self.end
    }

    pub fn len(&self) -> usize {
        self.end - self.start
    }

    pub fn is_empty(&self) -> bool {
        self.start == self.end
    }

    pub fn as_range(&self) -> Range<usize> {
        self.start..self.end
    }

    pub fn union(&self, other: Self) -> Self {
        Self {
            start: self.start.min(other.start),
            end: self.end.max(other.end),
        }
    }

    pub fn intersection(&self, other: Self) -> Option<Self> {
        let start = self.start.max(other.start);
        let end = self.end.min(other.end);

        if start < end {
            Some(Self { start, end })
        } else {
            None
        }
    }

    pub fn with_file(self, file_id: usize) -> FileSpan {
        FileSpan {
            span: self,
            file_id,
        }
    }
}

impl From<Span> for Range<usize> {
    fn from(span: Span) -> Self {
        span.as_range()
    }
}

impl std::fmt::Debug for Span {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.as_range().fmt(f)
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) struct FileSpan {
    pub span: Span,
    pub file_id: usize,
}

impl FileSpan {
    pub fn label_primary(self) -> Label<usize> {
        Label::primary(self.file_id, self.span)
    }

    pub fn label_secondary(self) -> Label<usize> {
        Label::secondary(self.file_id, self.span)
    }
}
