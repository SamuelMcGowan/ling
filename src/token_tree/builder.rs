use super::{TokenList, TokenTree};
use crate::diagnostic::{BracketMismatch, DiagnosticReporter};
use crate::lexer::token::{Bracket, BracketKind, Token, TokenKind};
use crate::lexer::Lexer;
use crate::source::Span;

pub(super) fn build_token_stream(lexer: Lexer, diagnostics: DiagnosticReporter) -> TokenList {
    TokenStreamBuilder::new(lexer, diagnostics).build()
}

struct SpareBracket {
    kind: BracketKind,
    token: Token,
}

struct TokenStreamBuilder<'a> {
    lexer: Lexer<'a>,

    diagnostics: DiagnosticReporter<'a>,
}

impl<'a> TokenStreamBuilder<'a> {
    fn new(lexer: Lexer<'a>, diagnostics: DiagnosticReporter<'a>) -> Self {
        Self { lexer, diagnostics }
    }

    fn build(mut self) -> TokenList {
        let mut trees = vec![];

        while let Some(token) = self.lexer.next() {
            let (tree, spare_bracket) = self.parse_tree(token);

            if let Some(tree) = tree {
                trees.push(tree);
            }

            // Nobody claimed this poor bracket :(
            if let Some(spare_bracket) = spare_bracket {
                // we don't emit a tree for unmatched closing brackets
                self.diagnostics.report(BracketMismatch {
                    bracket: Bracket::Closing(spare_bracket.kind),
                    span: spare_bracket.token.span,
                });
            }
        }

        TokenList(trees)
    }

    fn parse_tree(&mut self, token: Token) -> (Option<TokenTree>, Option<SpareBracket>) {
        match token.kind {
            TokenKind::Bracket(Bracket::Opening(kind)) => {
                // ooh look, a tree
                let (tokens, span, spare_bracket) = self.parse_group(token, kind);
                (
                    Some(TokenTree::Group {
                        bracket_kind: kind,
                        tokens,
                        span,
                    }),
                    spare_bracket,
                )
            }
            TokenKind::Bracket(Bracket::Closing(kind)) => {
                // an unexpected closing bracket! aargh!
                (None, Some(SpareBracket { kind, token }))
            }
            _ => (Some(TokenTree::Token(token)), None),
        }
    }

    fn parse_group(
        &mut self,
        token: Token,
        kind: BracketKind,
    ) -> (TokenList, Span, Option<SpareBracket>) {
        let mut trees = vec![];

        macro_rules! prev_span {
            () => {
                trees
                    .last()
                    .map(|tree: &TokenTree| tree.span())
                    .unwrap_or(token.span)
            };
        }

        let (end_span, spare_bracket) = loop {
            // get a token
            let Some(inner_token) = self.lexer.next() else {
                // unexpected end of input
                self.diagnostics.report(BracketMismatch {
                    bracket: Bracket::Opening(kind),
                    span: token.span,
                });
                break (prev_span!(), None);
            };

            // check for our closing bracket
            if matches!(inner_token.kind, TokenKind::Bracket(Bracket::Closing(k)) if k == kind) {
                break (inner_token.span, None);
            }

            // parse a child tree
            let (tree, spare_bracket) = self.parse_tree(inner_token);

            if let Some(tree) = tree {
                trees.push(tree);
            }

            // check the spare bracket
            if let Some(spare_bracket) = spare_bracket {
                if spare_bracket.kind == kind {
                    // our closing bracket was found!
                    break (spare_bracket.token.span, None);
                } else {
                    // not our bracket
                    // our bracket was unterminated too, so we report that
                    self.diagnostics.report(BracketMismatch {
                        bracket: Bracket::Opening(kind),
                        span: token.span,
                    });
                    break (prev_span!(), Some(spare_bracket));
                }
            }
        };

        let span = token.span.union(end_span);

        (TokenList(trees), span, spare_bracket)
    }
}
