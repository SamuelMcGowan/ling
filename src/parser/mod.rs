pub mod expr;
pub mod item;

use crate::ast::Spanned;
use crate::diagnostic::{DiagnosticReporter, ParseError, ParseResult};
use crate::lexer::token::{BracketKind, Token, TokenKind};
use crate::token_tree::{TokenIter, TokenTree};

pub(crate) struct Parser<'a> {
    tokens: TokenIter,
    diagnostics: DiagnosticReporter<'a>,
}

// Utilities.
impl<'a> Parser<'a> {
    pub fn new(tokens: TokenIter, diagnostics: DiagnosticReporter<'a>) -> Self {
        Self {
            tokens,
            diagnostics,
        }
    }

    fn parser_for_tokens(&mut self, tokens: TokenIter) -> Parser {
        Parser {
            tokens,
            diagnostics: self.diagnostics.borrow(),
        }
    }

    fn eat_kind(&mut self, kind: TokenKind) -> bool {
        match self.tokens.peek() {
            Some(TokenTree::Token(token)) if token.kind == kind => {
                self.tokens.next();
                true
            }
            _ => false,
        }
    }

    fn expect_kind(&mut self, kind: TokenKind) -> ParseResult<Token> {
        match self.tokens.next() {
            Some(TokenTree::Token(token)) if token.kind == kind => Ok(token),
            other => Err(self.unexpected(format!("token of kind {kind:?}"), other)),
        }
    }

    fn expect_group(
        &mut self,
        kind: BracketKind,
        expected: impl Into<String>,
    ) -> ParseResult<TokenIter> {
        match self.tokens.next() {
            Some(TokenTree::Group {
                bracket_kind,
                tokens,
                ..
            }) if bracket_kind == kind => Ok(tokens.into_iter()),
            other => Err(self.unexpected(expected, other)),
        }
    }

    fn expect_end(&mut self, tree_name: &str) -> ParseResult<()> {
        match self.tokens.next() {
            None => Ok(()),
            Some(tt) => Err(self.unexpected(format!("end of {tree_name}"), Some(tt))),
        }
    }

    fn eat_group(&mut self, kind: BracketKind) -> Option<TokenIter> {
        match self.tokens.peek() {
            Some(TokenTree::Group { bracket_kind, .. }) if *bracket_kind == kind => {
                let Some(TokenTree::Group { tokens, ..}) = self.tokens.next() else {
                    unreachable!();
                };

                Some(tokens.into_iter())
            }
            _ => None,
        }
    }

    fn parse_list<T>(
        &mut self,
        tree_name: &str,
        delim: TokenKind,
        parse: impl Fn(&mut Self) -> ParseResult<T>,
    ) -> ParseResult<Vec<T>> {
        let mut items = vec![];

        while !self.tokens.at_end() {
            let item = parse(self)?;
            items.push(item);

            if !self.eat_kind(delim) {
                if let Some(token) = self.tokens.next() {
                    return Err(self.unexpected(
                        format!("token of kind {delim:?} or end of {tree_name:?}"),
                        Some(token),
                    ));
                }
                break;
            }
        }

        Ok(items)
    }
}

// Error handling/recovery.
impl Parser<'_> {
    fn parse_or_recover<T>(
        &mut self,
        parser: impl Fn(&mut Self) -> ParseResult<T>,
        recover: impl Fn(&mut Self) -> T,
    ) -> T {
        parser(self).unwrap_or_else(|error| {
            self.diagnostics.report(error);
            recover(self)
        })
    }

    fn parse_spanned<T>(&mut self, parser: impl Fn(&mut Self) -> T) -> Spanned<T> {
        let start = self.tokens.peek().map(|tt| tt.span());
        let inner = parser(self);
        let end = self.tokens.prev_span();

        let span = match (start, end) {
            (Some(start), Some(end)) => Some(start.union(end)),
            _ => None,
        };

        Spanned { inner, span }
    }

    fn recover_until(&mut self, kinds: &[TokenKind]) {
        while let Some(tt) = self.tokens.peek() {
            if matches!(tt, TokenTree::Token(token) if kinds.contains(&token.kind)) {
                break;
            }
            self.tokens.next();
        }
    }

    fn recover_past(&mut self, kind: TokenKind) {
        self.tokens
            .by_ref()
            .take_while(|token| !matches!(token, TokenTree::Token(token) if token.kind == kind))
            .for_each(|_| {});
    }

    fn unexpected(&self, expected: impl Into<String>, tt: Option<TokenTree>) -> ParseError {
        let span = match &tt {
            Some(tt) => tt.span(),
            None => self.tokens.end_span(),
        };
        ParseError::unexpected(expected.into(), span)
    }
}

#[cfg(test)]
pub(crate) fn test_lex(
    source: &str,
) -> (
    crate::token_tree::TokenList,
    crate::diagnostic::DiagnosticOutput,
) {
    use crate::lexer::Lexer;
    use crate::source::with_test_source;
    use crate::token_tree::TokenList;

    with_test_source(source, |source, diagnostics| {
        let lexer = Lexer::new(source);
        TokenList::from_lexer(lexer, diagnostics)
    })
}

#[cfg(test)]
pub(crate) fn test_parse<T>(
    source: &str,
    f: impl Fn(&mut Parser) -> T,
) -> (T, crate::diagnostic::DiagnosticOutput) {
    use crate::lexer::Lexer;
    use crate::source::with_test_source;
    use crate::token_tree::TokenList;

    let (res, diagnostic_output) = with_test_source(source, |source, mut diagnostics| {
        let lexer = Lexer::new(source);
        let tokens = TokenList::from_lexer(lexer, diagnostics.borrow());

        let mut parser = Parser::new(tokens.into_iter(), diagnostics);
        f(&mut parser)
    });

    (res, diagnostic_output)
}
