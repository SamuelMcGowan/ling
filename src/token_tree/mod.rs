mod builder;

use std::vec::IntoIter;

use serde::Serialize;

use self::builder::build_token_stream;
use crate::diagnostic::DiagnosticReporter;
use crate::lexer::token::{BracketKind, Token};
use crate::lexer::Lexer;
use crate::source::span::Span;

#[derive(Debug, Clone, Serialize)]
pub(crate) enum TokenTree {
    Token(Token),
    Group {
        bracket_kind: BracketKind,
        tokens: TokenList,
        span: Span,
    },
}

impl TokenTree {
    pub fn span(&self) -> Span {
        match self {
            Self::Token(token) => token.span,
            Self::Group { span, .. } => *span,
        }
    }
}

#[derive(Debug, Clone)]
pub(crate) struct TokenList {
    trees: Vec<TokenTree>,
    end_span: Span,
}

impl TokenList {
    pub fn from_lexer(lexer: Lexer, diagnostics: DiagnosticReporter) -> Self {
        build_token_stream(lexer, diagnostics)
    }

    pub fn len(&self) -> usize {
        self.trees.len()
    }

    pub fn is_empty(&self) -> bool {
        self.trees.is_empty()
    }
}

impl IntoIterator for TokenList {
    type IntoIter = TokenIter;
    type Item = TokenTree;

    fn into_iter(self) -> Self::IntoIter {
        TokenIter {
            iter: self.trees.into_iter(),
            prev_span: None,
            end_span: self.end_span,
        }
    }
}

impl Serialize for TokenList {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        self.trees.serialize(serializer)
    }
}

pub(crate) struct TokenIter {
    iter: IntoIter<TokenTree>,
    prev_span: Option<Span>,
    end_span: Span,
}

impl TokenIter {
    pub fn at_end(&self) -> bool {
        self.peek().is_none()
    }

    pub fn peek(&self) -> Option<&TokenTree> {
        self.iter.as_slice().iter().next()
    }

    pub fn prev_span(&self) -> Option<Span> {
        self.prev_span
    }

    pub fn end_span(&self) -> Span {
        self.end_span
    }
}

impl Iterator for TokenIter {
    type Item = TokenTree;

    fn next(&mut self) -> Option<Self::Item> {
        let item = self.iter.next()?;
        self.prev_span = Some(item.span());
        Some(item)
    }

    fn size_hint(&self) -> (usize, Option<usize>) {
        self.iter.size_hint()
    }
}

impl ExactSizeIterator for TokenIter {}

#[cfg(test)]
mod tests {
    use insta::assert_ron_snapshot;

    use crate::parser::test_lex;

    #[test]
    fn simple_tests() {
        assert_ron_snapshot!("token_tree_0", test_lex("( { ) } ( hello )"));
        assert_ron_snapshot!("token_tree_1", test_lex("{{{}}}"));
        assert_ron_snapshot!("token_tree_2", test_lex("((("));
        assert_ron_snapshot!("token_tree_3", test_lex(")))"));
        assert_ron_snapshot!("token_tree_4", test_lex("]{}["));

        assert_ron_snapshot!(
            "token_tree_func",
            test_lex(
                "
            func foo(a: uint, b: uint) -> uint {
                a + b
            }
        "
            )
        );
    }
}
