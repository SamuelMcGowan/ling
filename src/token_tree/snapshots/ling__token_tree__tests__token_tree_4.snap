---
source: src/token_tree/mod.rs
expression: "test_lex(\"]{}[\")"
---
([
  Group(
    bracket_kind: Curly,
    tokens: [],
    span: Span(
      start: 1,
      end: 3,
    ),
  ),
  Group(
    bracket_kind: Square,
    tokens: [],
    span: Span(
      start: 3,
      end: 4,
    ),
  ),
], "error: mismatched Closing(Square)\n  ┌─ test_source:1:1\n  │\n1 │ ]{}[\n  │ ^\n\nerror: mismatched Opening(Square)\n  ┌─ test_source:1:4\n  │\n1 │ ]{}[\n  │    ^\n\n")
